File: version_1/models/scalar_ou_lqg/simulation.py
"""Simulator for the scalar OU LQG model — numpy reference.

Generates synthetic open-loop or closed-loop trajectories with truth
parameters. Used to produce data for the filter side (Stage A1) and
for cost evaluation in the control side (Stage A2/A3).
"""

from __future__ import annotations

from dataclasses import dataclass

import math
import numpy as np


# ── Parameter sets ─────────────────────────────────────────────────────

PARAM_SET_A = {
    'a':       1.0,
    'b':       1.0,
    'sigma_w': 0.3,
    'sigma_v': 0.2,
    'q':       1.0,
    'r':       0.1,
    's':       1.0,
}

INIT_STATE_A = {'x_0': 0.0}

EXOGENOUS_A = {
    'dt':  0.05,
    'T':   20,           # number of grid steps
    'x0_var': 1.0,       # initial state variance
}


# ── Drift + diffusion ─────────────────────────────────────────────────

def drift(t, y, params, aux):
    """numpy drift: dx/dt = -a*x + b*u(t)."""
    del t
    x = y[0]
    u_t = aux.get('u_at_t', 0.0)
    return np.array([-params['a'] * x + params['b'] * u_t])


def diffusion_diagonal(params):
    return np.array([params['sigma_w']])


# ── Observation channel ───────────────────────────────────────────────

def gen_obs(trajectory, t_grid, params, aux, prior_channels, seed):
    """Gaussian observation channel: y_k = x_k + N(0, sigma_v^2)."""
    del aux, prior_channels
    rng = np.random.default_rng(seed)
    T = len(t_grid)
    x = trajectory[:, 0]
    noise = rng.normal(0.0, params['sigma_v'], size=T)
    return {
        't_idx':     np.arange(T, dtype=np.int32),
        'obs_value': (x + noise).astype(np.float32),
    }


# ── Direct trajectory simulator (numpy, used by tests + benches) ──────

def simulate(
    *,
    params: dict,
    init_state: dict,
    exogenous: dict,
    u: np.ndarray | None = None,
    seed: int = 0,
) -> dict:
    """Simulate one trajectory of the scalar OU LQG model.

    Args:
        params:     dict with a, b, sigma_w, sigma_v
        init_state: dict with x_0
        exogenous:  dict with dt, T, x0_var
        u:          (T,) control schedule. If None, open-loop u=0.
        seed:       RNG seed.

    Returns dict with keys: t_grid (T,), trajectory (T, 1), obs (T,),
    u (T,).
    """
    dt = float(exogenous['dt'])
    T = int(exogenous['T'])
    x0_var = float(exogenous.get('x0_var', 1.0))
    x0_mean = float(init_state['x_0'])

    rng = np.random.default_rng(seed)
    A = 1.0 - params['a'] * dt
    B = params['b'] * dt
    sw = params['sigma_w'] * math.sqrt(dt)

    if u is None:
        u = np.zeros(T)
    u = np.asarray(u, dtype=np.float64)

    x = np.zeros(T)
    x[0] = x0_mean + math.sqrt(x0_var) * rng.standard_normal()
    for k in range(1, T):
        x[k] = A * x[k - 1] + B * u[k - 1] + sw * rng.standard_normal()

    obs = x + params['sigma_v'] * rng.standard_normal(size=T)

    return {
        't_grid':     np.arange(T) * dt,
        'trajectory': x[:, None].astype(np.float64),
        'obs':        obs.astype(np.float64),
        'u':          u,
    }

File: version_1_Julia/models/scalar_ou_lqg/simulation.jl
# Simulator for the scalar OU LQG model — Julia reference.
# Two paths:
#   1. `simulate_em` — hand-rolled Euler-Maruyama, line-for-line port
#      of `version_1/models/scalar_ou_lqg/simulation.py:simulate`.
#      Used for bit-equivalence checks against the Python reference.
#   2. `simulate_diffeq` — `StochasticDiffEq.jl` (production path).
#      Same EM solver at the same dt → numerically equivalent to (1).

module Simulation

using Random: AbstractRNG, MersenneTwister
using StochasticDiffEq: SDEProblem, EM, solve

# ── Parameter sets ──────────────────────────────────────────────────────────
# Match the Python `PARAM_SET_A` / `INIT_STATE_A` / `EXOGENOUS_A` exactly.

const PARAM_SET_A = (
    a       = 1.0,
    b       = 1.0,
    sigma_w = 0.3,
    sigma_v = 0.2,
    q       = 1.0,
    r       = 0.1,
    s       = 1.0,
)

const INIT_STATE_A = (x_0 = 0.0,)

const EXOGENOUS_A = (
    dt     = 0.05,
    T      = 20,
    x0_var = 1.0,
)

# ── Path 1: hand-rolled Euler-Maruyama (matches Python `simulate`) ──────────

"""
    simulate_em(; params, init_state, exogenous, u=nothing, seed=0)

Direct port of `simulation.py:simulate`. The recurrence
    x_{k+1} = A·x_k + B·u_k + sw·ξ_k,  ξ_k ~ N(0, 1)
with A = 1 - a·dt, B = b·dt, sw = σ_w·√dt — bit-equivalent to Python at the
same RNG seed (modulo Python's `np.random` vs Julia's `MersenneTwister`).

Returns NamedTuple `(t_grid, trajectory, obs, u)` matching the Python dict.
"""
function simulate_em(; params = PARAM_SET_A,
                       init_state = INIT_STATE_A,
                       exogenous  = EXOGENOUS_A,
                       u::Union{Nothing,AbstractVector} = nothing,
                       seed::Integer = 0)
    dt     = exogenous.dt
    T      = exogenous.T
    x0_var = exogenous.x0_var
    x0_mean = init_state.x_0

    A  = 1.0 - params.a * dt
    B  = params.b * dt
    sw = params.sigma_w * sqrt(dt)

    rng = MersenneTwister(seed)
    u_vec = u === nothing ? zeros(Float64, T) : Float64.(u)

    x = zeros(Float64, T)
    x[1] = x0_mean + sqrt(x0_var) * randn(rng)
    for k in 2:T
        x[k] = A * x[k-1] + B * u_vec[k-1] + sw * randn(rng)
    end
    obs = x .+ params.sigma_v .* randn(rng, T)

    return (
        t_grid     = collect(0:T-1) .* dt,
        trajectory = reshape(x, :, 1),
        obs        = obs,
        u          = u_vec,
    )
end


# ── Path 2: StochasticDiffEq.jl (production path) ───────────────────────────

"""
    simulate_diffeq(; params, init_state, exogenous, u=nothing, seed=0)

Same model, same dt, same EM integrator — but routed through
`StochasticDiffEq.jl`'s `EM()` solver. Used to confirm the Julia
ecosystem's path agrees with the hand-rolled one and to demonstrate the
production interface (charter §13: `sde_solver_diffrax.py` →
`StochasticDiffEq.jl`).

Note: StochasticDiffEq's `EM` consumes a `(du, u, p, t)` in-place drift
and a separate `(du, u, p, t)` diffusion (g) callback. The Wiener-noise
RNG is internal to StochasticDiffEq and not bit-comparable with the
hand-rolled MersenneTwister path; numerical agreement is statistical
(across seeds), not bit-for-bit.
"""
function simulate_diffeq(; params = PARAM_SET_A,
                           init_state = INIT_STATE_A,
                           exogenous  = EXOGENOUS_A,
                           u::Union{Nothing,AbstractVector} = nothing,
                           seed::Integer = 0)
    dt = exogenous.dt
    T  = exogenous.T
    x0_mean = init_state.x_0
    x0_var  = exogenous.x0_var
    u_vec   = u === nothing ? zeros(Float64, T) : Float64.(u)

    # In-place drift / diffusion. We bake u(t) as a step-wise function of t.
    function f!(du, x, p, t)
        # control schedule: piecewise-constant at floor(t/dt) bin
        k     = clamp(Int(floor(t / dt)) + 1, 1, T)
        u_t   = u_vec[k]
        du[1] = -params.a * x[1] + params.b * u_t
        return nothing
    end
    function g!(du, x, p, t)
        du[1] = params.sigma_w
        return nothing
    end

    rng_init = MersenneTwister(seed)
    x0 = [x0_mean + sqrt(x0_var) * randn(rng_init)]

    prob = SDEProblem(f!, g!, x0, (0.0, (T - 1) * dt), nothing)
    sol  = solve(prob, EM(); dt = dt, saveat = 0:dt:(T-1)*dt, seed = seed)

    traj = reduce(hcat, sol.u)'   # (T, 1)
    # Observations sampled with a fresh RNG so seed reuses the Wiener stream
    # without cross-contaminating the obs noise.
    rng_obs = MersenneTwister(seed + 1_000_003)
    obs = vec(traj[:, 1]) .+ params.sigma_v .* randn(rng_obs, T)

    return (
        t_grid     = collect(0:T-1) .* dt,
        trajectory = traj,
        obs        = obs,
        u          = u_vec,
    )
end

end # module Simulation
